text/0000-classic-decorator.md

```diff
@@ -0,0 +1,196 @@
+   child class.
+
+This should allow users who wish to adopt native class syntax to do so
+incrementally, one change at a time, with as much safety as possible.
+
+## Detailed design
+
+There are roughly two categories of base classes in Ember:
+
+1. Base classes which do not have alternatives, _and_ use a very minimal amount
+   of `EmberObject` APIs
+   - `Route`
+   - `Controller`
+   - `Service`
+   - `Helper`
+2. Base classes which have modern alternatives, _and_ use a much larger amount
+   of `EmberObject` APIs:
+   - `Component`, which can be converted to `GlimmerComponent`
+   - Utility classes, which can be converted plain native classes that do not
+     extend `EmberObject`
+
+`@classic` will work slightly differently for these two categories. We'll call
+them _transitionable_ and _non-transitionable_ respectively.
+
+### Transitionable Base Classes
+
+Transitionable base classes can be converted to native classes safely _without_
+using any `EmberObject` APIs. Once converted, we can safely lint against using
+any `EmberObject` APIs, preventing any future confusion. As such, it will only
+be necessary to apply the `@classic` decorator for as long as any of the user's
+own classes use `EmberObject` APIs. These include:
+
+- `extend`
+- `reopen` and `reopenClass`
+- `init`
+- `destroy`
+- `proto`
+- `detect`
+- `get`/`set`
+
+Along with any other extra methods or APIs that exist on `CoreObject` and
+`EmberObject`. Methods and APIs that exist on the class itself, such as
+`transitionTo` on Routes, or `send` and `sendAction` on Controllers, are _not_
+considered `EmberObject` APIs, since they could be implemented purely in native
+classes, and aren't tied to the way `EmberObject` does inheritance.
+
+If a class, or any of its parent classes, uses one of these methods _without_
+being decorated with `@classic`, a warning will be logged to let users know that
+they are using classic object APIs, and they should either refactor away from
+using those APIs, or use the `@classic` decorator to opt-out of warnings.
+
+The `@classic` decorator itself will brand the class it is applied to with a
+symbol, disabling warnings on that instance of the class.
+
+### Non-transitionable Base Classes
+
+Non-transitionable base classes _cannot_ be converted to native classes without
+relying on behaviors that are specific to `EmberObject`. Classic components, for
+instance, rely deeply on the way that `EmberObject.create` sets up its instance,
+and any utility class that extends from `EmberObject` _must_ use `create` to
+create instances. While this is also true of `Route`s and `Service`s, the
+important distinction is that _users_ don't have to ever use `create` in their
+own code, or are not passed any arguments other than injections, so for those
+classes the usage of `create` is an implementation detail.
+
+These classes will _always_ warn users if they are not decorated with
```

Comment:
is this a linter warning or a deprecation style warning?

Reply:
This would be a deprecation style warning, since it’s not information we can actually gather through static analysis. We need to actually run/construct the class to get this information

Reply:
Will it actually be a deprecation, with an identifier and an `until` field? Or just use warn from 'ember-debug'?